src/back/glsl/mod.rs

```diff
@@ -2961,25 +2940,309 @@ impl<'a, W: Write> Writer<'a, W> {
+            if let Some(level_expr) = level {
+                self.write_expr(level_expr, ctx)?;
+                write!(self.out, " < textureQueryLevels(",)?;
+                self.write_expr(image, ctx)?;
+                // Chain the next check
+                write!(self.out, ") && ")?;
+            }
+
+            // Check that the sample arguments doesn't exceed the number of samples
+            if let Some(sample_expr) = sample {
+                self.write_expr(sample_expr, ctx)?;
+                write!(self.out, " < textureSamples(",)?;
+                self.write_expr(image, ctx)?;
+                // Chain the next check
+                write!(self.out, ") && ")?;
+            }
+
+            // We now need to write the size checks for the coordinates and array index
+            // first we write the comparation function in case the image is 1D non arrayed
+            // (and no 1D to 2D hack was needed) we are comparing scalars so the less than
+            // operator will suffice, but otherwise we'll be comparing two vectors so we'll
+            // need to use the `lessThan` function but it returns a vector of booleans (one
+            // for each comparison) so we need to fold it all in one scalar boolean, since
+            // we want all comparisons to pass we use the `all` function which will only
+            // return `true` if all the elements of the boolean vector are also `true`.
+            //
+            // So we'll end with one of the following forms
+            // - `coord < textureSize(image, lod)` for 1D images
+            // - `all(lessThan(coord, textureSize(image, lod)))` for normal images
+            // - `all(lessThan(ivec(coord, array_index), textureSize(image, lod)))`
+            //    for arrayed images
+            // - `all(lessThan(coord, textureSize(image)))` for multi sampled images
+
+            if vector_size != 1 {
+                write!(self.out, "all(lessThan(")?;
+            }
+
+            // If the image needs an array indice we need to add it to the end of our
+            // coordinate vector, to do so we will use the `ivec(ivec, scalar)`
+            // constructor notation (NOTE: the inner `ivec` can also be a scalar, this
+            // is important for 1D arrayed images).
+            if array_index.is_some() {
+                write!(self.out, "ivec{}(", vector_size)?;
+            }
+            // Write the coordinates always in the middle of the constructor
+            // as the first argument (in case the constructor is needed)
+            self.write_expr(coordinate, ctx)?;
+            // Write the array index after the coordinates
+            if let Some(array_expr) = array_index {
+                // Separate the coordinates from the index
+                write!(self.out, ", ")?;
+                self.write_expr(array_expr, ctx)?;
+                // Close the vector constructor
+                write!(self.out, ")")?;
+            }
```

Comment:
It seems like we have very similar code in `write_image_store` and `write_image_load` (several times) for generating these extended coordinate expressions, adding in the 1D hack and the array layer. Could we pull this out into a `write_texture_coord` helper function that takes a `tex_1d_hack: bool` argument, and just pass that as `false` in this case, since we're handling the 1D case separately?